11_FFTBloom/app_resources/kernel_fft_first_axis.hlsl

```diff
@@ -0,0 +1,154 @@
+{
+	
+	void preload(uint32_t channel)
+	{
+		float32_t2 inputImageSize;
+		texture.GetDimensions(inputImageSize.x, inputImageSize.y);
+		float32_t2 normalizedCoordsFirstLine, normalizedCoordsSecondLine;
+		normalizedCoordsFirstLine.x = (float32_t(2 * glsl::gl_WorkGroupID().x)+0.5f)/(inputImageSize.x*KERNEL_SCALE);
+		normalizedCoordsSecondLine.x = (float32_t(2 * glsl::gl_WorkGroupID().x + 1)+0.5f)/(inputImageSize.x*KERNEL_SCALE);
+
+		for (uint32_t localElementIndex = 0; localElementIndex < ELEMENTS_PER_THREAD; localElementIndex++)
+		{
+			// Index computation here is easier than FFT since the stride is fixed at _NBL_HLSL_WORKGROUP_SIZE_
+			const uint32_t index = _NBL_HLSL_WORKGROUP_SIZE_ * localElementIndex | workgroup::SubgroupContiguousIndex();
+			normalizedCoordsFirstLine.y = (float32_t(index) + 0.5f) / (inputImageSize.y * KERNEL_SCALE);
+			normalizedCoordsSecondLine.y = normalizedCoordsFirstLine.y;
+			preloaded[localElementIndex].real(scalar_t(texture.SampleLevel(samplerState, normalizedCoordsFirstLine + promote<float32_t2, float32_t>(0.5 - 0.5 / KERNEL_SCALE), 0)[channel]));
+			preloaded[localElementIndex].imag(scalar_t(texture.SampleLevel(samplerState, normalizedCoordsSecondLine + promote<float32_t2, float32_t>(0.5 - 0.5/KERNEL_SCALE), 0)[channel]));
+		}
+
+		// Set LegacyBdaAccessor for posterior writing
+		colMajorAccessor = LegacyBdaAccessor<complex_t<scalar_t> >::create(getChannelStartAddress(channel));
+	}
+
+	// Util to write values to output buffer in column major order - this ensures coalesced writes
+	void storeColMajor(uint32_t index, NBL_CONST_REF_ARG(complex_t<scalar_t>) firstValue, NBL_CONST_REF_ARG(complex_t<scalar_t>) secondValue)
+	{
+		colMajorAccessor.set(colMajorOffset(2 * glsl::gl_WorkGroupID().x, index), firstValue);
+		colMajorAccessor.set(colMajorOffset(2 * glsl::gl_WorkGroupID().x + 1, index), secondValue);
+	}
+
+	// Once the FFT is done, each thread should write its elements back. Storage is in column-major order because this avoids cache misses when writing.
+	// Channels will be contiguous in buffer memory. We only need to store outputs 0 through Nyquist, since the rest can be recovered via complex conjugation:
+	// see https://en.wikipedia.org/wiki/Discrete_Fourier_transform#DFT_of_real_and_purely_imaginary_signals
+	// FFT unpacking rules explained in the readme
+	// Also, elements 0 and Nyquist fit into a single complex element since they're both real, and it's always thread 0 holding these values
+	template<typename SharedmemAdaptor>
+	void unload(uint32_t channel, SharedmemAdaptor sharedmemAdaptor)
+	{
+		// Storing even elements of NFFT is storing the bitreversed lower half of DFT - see readme
+		for (uint32_t localElementIndex = 0; localElementIndex < ELEMENTS_PER_THREAD; localElementIndex += 2)
+		{
+			// First element of 0th thread has a special packing rule
+			if (!workgroup::SubgroupContiguousIndex() && !localElementIndex)
```

Comment:
in HLSL like C++ boolean operators in short-circuit, you want to do the unrollable constexpr (dynamically uniform) first, this way it will actually get optimized out during loop unroll